llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3409,6 +3409,25 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
       return replaceOperand(SI, 2, S);
   }
 
+  {
+    // A pred C ? (A >> BW - 1) : 1 --> ZExt(A pred C ? A < 0 : 1)
```

Comment:
Should be generalized to `Cond ? (A >> (BW - 1)) : 1 --> ZExt(Cond ? A < 0 : 1)`.

I am sorry I cannot give the approval for this patch if you cannot demonstrate that it benefits some real-world applications.

Reply:
AFAICT the idea is that this is useful if `A pred C` implies something about `A < 0` in which case it makes sense to fold `A >> BW - 1` -> the more easy to reason about `A < 0`.
At least thats what it looks like the purpose is from the tests.

Reply:
@goldsteinn, Should I fold `A < 0` instead of `A Pred C`? or draft the patch until we can demonstrate that it benefits some real-world applications

Reply:
No real comment about the real-world impact. I'm generally pretty in favor of "get anything we can in" but think thats a minority opinion.

I would refactor to make the goal explicit. Use `makeExactICmpRegion` on the `A Pred C` and just do the complete simplification if the range simplifies `A < 0` to a constant.